### PR TITLE
Go 1.19 support (libp2p upgrade)

### PR DESCRIPTION
**Pull request by SaveTheRbtz**
* https://github.com/libp2p/go-libp2p-swarm is deprecated.
* https://github.com/libp2p/go-libp2p-discovery is deprecated.
* https://github.com/libp2p/go-tcp-transport is deprecated.
* `blockstore.ErrNotFound` is no longer exported exported and replaced with `ipld.IsNotFound(err)`.
* `go vet` [now complains about wrong usage](https://tip.golang.org/doc/go1.19#vet) of `errors.As` which we have plenty 

As a tiny bonus, a 3% perf improvement based on the ComputeBlock benchmark:
```
$ benchstat old new   
name                             old time/op    new time/op    delta
ComputeBlock/16/cols/128/txes-8     2.09s ± 1%     2.03s ± 2%  -2.99%  (p=0.000 n=9+10)

name                             old us/tx      new us/tx      delta
ComputeBlock/16/cols/128/txes-8     1.02k ± 1%     0.99k ± 2%  -2.98%  (p=0.000 n=9+10)

name                             old alloc/op   new alloc/op   delta
ComputeBlock/16/cols/128/txes-8    1.88GB ± 1%    1.87GB ± 0%    ~     (p=0.370 n=9+8)

name                             old allocs/op  new allocs/op  delta
ComputeBlock/16/cols/128/txes-8     29.3M ± 0%     29.2M ± 0%  -0.13%  (p=0.000 n=8+8)
```

### CODE REVIEW

**Comment by SaveTheRbtz** [comment]
@tarakby seems like `github.com/btcsuite/btcd` is no longer developed and crypto would need to migrate to `github.com/btcsuite/btcd/btcec/v2` at some point.

**Comment by tarakby** [comment]
>  seems like github.com/btcsuite/btcd is no longer developed and ./crypto/ would need to migrate to github.com/btcsuite/btcd/btcec/v2 at some point.

There is a community PR with the upgrade, I'll follow up on that.

File: network/p2p/dht_test.go
Comment (pattyshack) [inline network/p2p/dht_test.go:149]: undo?

**Review by pattyshack — APPROVED** [review]

File: network/p2p/dht_test.go
Comment (SaveTheRbtz) [reply network/p2p/dht_test.go:149]: `n` is passed into the goroutine down the line so we need a copy of it.  It was always a bug, but it started racing more consistently with the new versions of libp2p libraries.

File: network/p2p/keyutils/keyTranslator_test.go
Comment (SaveTheRbtz) [inline network/p2p/keyutils/keyTranslator_test.go:9]: @tarakby this should be safe, but just in case tagging you.

File: network/p2p/dht_test.go
Comment (pattyshack) [reply network/p2p/dht_test.go:149]: you're passing s into the goroutine.  s != n

Comment (SaveTheRbtz) [reply network/p2p/dht_test.go:149]: n is used inside the function as:
```
ch <- n.Host().ID()
```

Comment (AlexHentschel) [reply network/p2p/dht_test.go:149]: nice catch!

From my perspective, it would be cleaner to avoid variable shadowing and write 
```golang
subReader := func(s *pubsub.Subscription, node *p2p.Node) {
  msg, err := s.Next(ctx)
  require.NoError(t, err)
  require.NotNil(t, msg)
  assert.Equal(t, data, msg.Data)
  ch <- node.Host().ID()
}
```
and then write in [line 165](https://github.com/onflow/flow-go/blob/b22e0a7c052b75b68659028e32c253fd2dd4cca3/network/p2p/dht_test.go#L165) 
```golang
go subReader(s, n)
```

File: network/p2p/compressed/compressedStream.go
Comment (AlexHentschel) [inline network/p2p/compressed/compressedStream.go:77]: Just to confirm, this means that we are swallowing any errors the this call might return. Are we sure that such errors are not of any interest (even for debugging purposes)? 
The `multierr` seemed a generally applicable solution, just wondering what your reasoning is for using a less-general approach.

**Review by AlexHentschel — APPROVED** [review]
Looks good. Thanks for cleaning up the error handling. 
Don't know about your comment [here](https://github.com/onflow/flow-go/pull/2988#discussion_r948201994), so we might want to wait for Tarak's response before merging.

File: network/p2p/compressed/compressedStream.go
Comment (SaveTheRbtz) [reply network/p2p/compressed/compressedStream.go:77]: cc: @yhassanzadeh13

Comment (yhassanzadeh13) [reply network/p2p/compressed/compressedStream.go:77]: @AlexHentschel as far as I investigated, closing the libp2p stream also closes the writer on it. However, here the `gzip` writer is a wrapper around the original writer of the libp2p stream. This poses a cyclic causality:
1. We close `gzip` writer (i.e., `c.w.Close()`) _prior_ to closing the stream (this was what the previous `multierr` was doing: `multierr.Combine(c.w.Close(), c.Stream.Close())`) . Then, with the new upgrade, once the `gzip` writer is closing, it also closes the underlying writer of the libp2p stream, which leaves the stream on a half-open status. Then when the original stream is closing (i.e., `c.Stream.Close()`) it fails to close as closing a half-open stream results in a `stream rest`, so it returns an error. 
2. We close `gzip` writer _after_ closing the original stream. Then `gzip` also tries closing the underlying writer of the libp2p stream. But since it is already closed it results in the closing an already closed stream, which again returns a `stream rest` error.

Hence, what matters is to first safely and gracefully close the libp2p stream (which also tells the remote node to free up its resource), and then invoke closing the `gzip` writer that was established on top of the libp2p stream's writer. Closing this `gzip` writer is with the goal of just cleaning up the resource locally making sure that there is no orphan resources lingering post stream closing. However, we are aware that closing that stream will cause a s `stream reset` error. 

To clarify this further, I added an error checking to check the returned error of closing `gzip` writer, and return it if it is not a `stream reset` error: https://github.com/onflow/flow-go/pull/3008

File: network/p2p/keyutils/keyTranslator_test.go
Comment (tarakby) [reply network/p2p/keyutils/keyTranslator_test.go:9]: Thanks for the flag. I had a look and all looks good 👌🏼

**Comment by SaveTheRbtz** [comment]
bors merge

**Comment by SaveTheRbtz** [comment]
> @tarakby seems like `github.com/btcsuite/btcd` is no longer developed and `./crypto/` would need to migrate to `github.com/btcsuite/btcd/btcec/v2` at some point.

Got an ack from Tarak on slack for the `btcd` changes.

**Comment by SaveTheRbtz** [comment]
bors retry

**Comment by SaveTheRbtz** [comment]
bors merge

**Comment by SaveTheRbtz** [comment]
bors merge

**Comment by SaveTheRbtz** [comment]
bors cancel

**Comment by SaveTheRbtz** [comment]
bors merge

**Comment by SaveTheRbtz** [comment]
bors merge

**Comment by SaveTheRbtz** [comment]
bors merge

**Comment by SaveTheRbtz** [comment]
bors try

**Comment by SaveTheRbtz** [comment]
bors merge p=-1

**Pull request merged**
